File: examples/go/post-quantum-slhdsa-naive-INSECURE/PostQuantumSLHDSANaiveInsecure_test.go
package contract

import (
	"testing"

	runar "github.com/icellan/runar/packages/runar-go"
)

// PostQuantumSLHDSANaiveInsecure.runar.go is a pedagogical artifact that shows
// the broken pattern of verifying a free post-quantum signature against a free
// message — anyone observing one valid spend can reuse the (msg, sig) pair (or
// substitute any other pair they hold) and the script still verifies. The
// correct hybrid pattern lives in examples/go/sphincs-wallet.
//
// Insecure is not the same as vacuous. The port used to carry `//go:build
// ignore` and this suite could only run the Rúnar frontend; the tag was there
// because the file imported the module path `"runar"`, which resolves to
// nothing. With the import corrected the contract builds as Go, and the row
// below is what separates real FIPS 205 verification from a stub.

func TestPostQuantumSLHDSANaiveInsecure_Compile(t *testing.T) {
	if err := runar.CompileCheck("PostQuantumSLHDSANaiveInsecure.runar.go"); err != nil {
		t.Fatalf("Rúnar compile check failed: %v", err)
	}
}

func TestPostQuantumSLHDSANaiveInsecure_RefusesGarbageSignature(t *testing.T) {
	defer func() {
		if recover() == nil {
			t.Fatal("a garbage signature was ACCEPTED — the SLH-DSA verifier " +
				"is not verifying anything")
		}
	}()
	c := &PostQuantumSLHDSANaiveInsecure{Pubkey: runar.ByteString(make([]byte, 32))}
	c.Spend(runar.ByteString("message"), runar.ByteString(make([]byte, 64)))
}

File: examples/rust/post-quantum-slhdsa-naive-INSECURE/PostQuantumSLHDSANaiveInsecure_test.rs
// PostQuantumSLHDSANaiveInsecure.runar.rs is a pedagogical artifact that
// shows the broken pattern of verifying a free post-quantum signature
// against a free message — anyone observing one valid spend can reuse the
// (msg, sig) pair (or substitute any other (msg, sig) they have) and the
// script still verifies. The correct hybrid pattern lives in
// examples/rust/sphincs-wallet.
//
// We could in principle import the contract via `#[path]` and demonstrate
// the broken behaviour with a native call (the verify_slh_dsa_sha2_128s
// SDK shim returns true under mocked crypto), but the SDK mock would
// trivially pass for arbitrary (msg, sig) inputs and obscure the real
// attack vector — which only manifests on-chain against a real verifier.
// We exercise the Rúnar frontend directly via compile_check, which is the
// cross-compiler conformance boundary we care about.

#[test]
fn test_compile() {
    runar::compile_check(
        include_str!("PostQuantumSLHDSANaiveInsecure.runar.rs"),
        "PostQuantumSLHDSANaiveInsecure.runar.rs",
    )
    .unwrap();
}
